src/utils/commands/command_util.py

```diff
@@ -109,82 +109,64 @@ def rank(self, discord_id: str) -> str:
-        `submissions` leetcode API to determine which of the weekly
-        challenge problems the given user has solved. Gathers the list
+        weekly challenge for the given leetcode_user_id. Calls update_user_completions to check
+        database for completions and updates via leetcode API ONLY if necessary. Gathers the list
         of current weekly challenges from the database.
         """
         result = ""
         user = self.database.users.load_by_discord_id(discord_id)
         if user:
-            leetcode_user_id = user["leetcode_id"]
             challenge = self.database.weekly_challenges.get_latest()
-            if not challenge:
-                result += "No current challenges"
-            else:
+            if challenge:
+                user_leetcode_id = user["leetcode_id"]
+                self.update_user_completions(user_leetcode_id, challenge["id"])
+                completions = self.database.question_completions.load_all_title_slugs_by_user(
+                    user_leetcode_id
+                    )
                 questions = self.database.weekly_questions.load_by_challenge_id(challenge["id"])
-                total = len(questions)
-                if total == 0:
-                    result += "No current weekly questions"
-                else:
-                    completed = 0
-                    for question in questions:
-                        question["complete"] = self.leetcode.check_challenge_completion(
-                            leetcode_user_id, question["title_slug"]
-                        )
-                        if question["complete"]:
-                            completed += 1
-                    percentage = int(completed/total * 100)
-                    result += f"User {leetcode_user_id}'s Weekly Challenge status: {percentage}%\n"
-                    for question in questions:
-                        complete = question["complete"]
-                        result += (
-                            f"\t{self.reaction_complete if complete else self.reaction_incomplete}"
-                            f"\t-\t{question['title']}\n"
+                question_slugs = [q['title_slug'] for q in questions]
+                completed_in_database = list(set(completions) & set(question_slugs))
+                total_completions = len(completed_in_database)
+                total_questions = len(question_slugs)
+                percentage = int(total_completions / total_questions * 100)
+                result += f"User {user_leetcode_id}'s Weekly Challenge status: {percentage}%\n"
+                completed_all = total_completions == total_questions
+                for question in questions:
+                    complete = True if completed_all else \
+                    self.database.question_completions.check_completion(
+                        user_leetcode_id,
+                        question["title_slug"]
                         )
```

Comment:
Earlier in the method we already have `completions` which should be all QuestionCompletion entries for the user. Can't we just check here to see `if question["title_slug"] in completions:` instead of making a conditional database call?

Reply:
That's a good point, made this change. Though I'm wondering if it's worth still including the O(1) check of `completed_all` and to only use `question["title_slug"] in completions` if completed_all is False? completions will never be a large list to iterate through, but still.